### ISSUE TITLE
Handle smart components

### ISSUE THREAD

**Issue by jamesgk** [googlefonts/glyphsLib#91]
Account for these when exporting UFOs from Glyphs source:
https://glyphsapp.com/tutorials/smart-components

**Comment by jamesgk** [comment]
Actually, it would help to see some practical example Glyphs source using smart components.
cc @m4rc1e who brought this up in https://github.com/googlei18n/glyphsLib/issues/64#issuecomment-237012031.

**Comment by m4rc1e** [comment]
Hey @jamesgk,

We amended an insane one recently, https://github.com/phoikoi/VT323. It's a pixel font made out of smart components.
![screen shot 2016-12-23 at 22 57 15](https://cloud.githubusercontent.com/assets/7525512/21462987/2f76c2c8-c963-11e6-9ae4-bac101ad9255.png)

I'll also see if I can find a more traditional font which uses smart components for serifs etc.

cc @fontHausen

**Comment by punchcutter** [comment]
At the moment glyphsLib will read smart components as if they were regular components without any of the smartness. It would be nice to have a warning because right now there is no error or warning and what gets generated is not what Glyphs outputs. As far as I can tell we just need to check if the component has a 'piece' attribute. If not then it's either not a smart component or the settings are at the default and can be used as is. If it does have 'piece' then there could be a warning indicating that glyphsLib doesn't support this yet.

**Comment by marekjez86** [comment]
@anthrotype : Cosimo what's the plan for smart components?

**Comment by anthrotype** [comment]
It's on my to-do list. It would great if somebody with greater familiarity with Glyphs.app than me could help figuring out the exact algorithm that Glyphs.app uses to make such "smart components". That is: figure out what are the pieces of data involved in this, where they are stored in the *.glyphs source, what each does etc., and describe how to go from X to Y. In plain English, not in code (the latter part is the easiest).

**Comment by punchcutter** [comment]
My simple explanation to people has been that basically every smart component has its own designspace with masters and axes. And "instances" are wherever the components are used so instead of interpolating each glyph in a font we need to interpolate components in each glyph. In theory it doesn't sound so bad to me because we are already doing this for font masters --> font instances. Now it just needs to happen in a mini way for smart component glyph masters --> composite glyph instances. 
@anthrotype soon you'll have the Noto Marchen Glyphs file with tons of smart components and around 800 glyphs so there's plenty to play with there.

**Comment by schriftgestalt** [comment]
For most smart components it will work just fine when you use Mutator math to do the interpolation. But there are a lot problems. Because oaf those we needed to adapt the master interpolation algorithm to be fast and capable enough to work for smart components, too. 
The new algorithm handles intermediate master much better so that is possible now, too.

**Comment by belluzj** [comment]
@anthrotype If that can help you, I can write a unit test with a smart component that is a rectangle parameterized with width/height, and a few instances that are easy to assert by comparing the coordinates of the four corners of the component.

I think I will have to write such a test anyway for the roundtrip so I can do it now if that's helpful.

**Comment by anthrotype** [comment]
that would be great, thanks guys!

**Comment by belluzj** [comment]
@anthrotype Here is a test: https://github.com/belluzj/glyphsLib/commit/9c03863f5d1d20ab2c8975393d90fd8233c73dc6

Here is the Glyphs file that the test builds, you can open it in Glyphs and play with the sliders to see how it interpolates: https://gist.github.com/belluzj/b0ecf98ff6eb8ed220365ad92c7b5bd5

![screen shot 2017-12-22 at 16 48 49](https://user-images.githubusercontent.com/3616772/34305877-2ae3c938-e738-11e7-9ebf-6be3911ff803.png)

**Comment by typoman** [comment]
It's been a long time, but it would be very nice to make smart components work in this package. Would there be any update on this?

**Comment by madig** [comment]
Nothing but thoughts so far. The problem is what to do with smart components when storing a UFO. 

- Do you just decompose, i.e. as Glyphs does it? This is the easiest way, but loses the semantics in case you want to roundtrip back.
- Do you decompose but mark the points, so glyphsLib "extracts" and reinserts the smart components? What if you modify the marked smart component outlines and roundtrip back to Glyphs, make a new one?
- Do you decompose and store the original stuff in a layer called whatever + `.smartComponents`? Again, what to do about modifications? Otherwise, smart components become a read-only concept. Not necessarily bad, but something to think about.
- Do you do something else?

And after this has been discussed, someone still needs to go and implement it. It's on my todo list, but has not been a priority so far.

**Comment by typoman** [comment]
I think the most straightforward solution, for now, is to decompose them on its location in every glyph. Idk about keeping them editable since they're not part of any other workflow other than the Glyphs app. If it was implemented in UFO then it would have been more convenient.

**Comment by gorjious** [comment]
It would be really helpful to see this implemented. Currently at a standstill, as I make use of Smart Components extensively. I agree the simplest solution would be to decompose each component with its set parameters. 
@chrissimpkins

**Comment by belluzj** [comment]
In case someone wants to work on this, I rebased the unit test I wrote three years ago onto the current master. It's a rectangle smart component with 3 axes.

**Comment by gorjious** [comment]
If it would help progress the issue, I would be happy to provide test files with examples of multiple parameters and nested smart components.

**Comment by anthrotype** [comment]
To reply to @madig's comment earlier https://github.com/googlefonts/glyphsLib/issues/91#issuecomment-547316790, maybe the best solution, that ensures smart components can be roundtripped back to .glyphs source and that potentially could allow UFO editors and workflows to be adapted to this feature, would be to export these smart components to some private data structure in the UFO and GLIF lib as plist format, then write a ufo2ft preProcessor filter that reads from that and decomposes the smart components upon building the binaries.

**Comment by marekjez86** [comment]
@madig : what do you think about the above @anthrotype comment/suggested solution?

**Comment by madig** [comment]
You mean use UFO components as placeholders and store the smartness in a glyphsLib-scoped lib key? Sounds good enough for now, but I won't have time any time soon to implement it.

**Comment by aaronbell** [comment]
I'm trying to prep https://github.com/matsuba723/Aoboshi for release on Google Fonts and unfortunately, the entire font is built with Glyphs' corner components, which it looks like glyphsLib / fontmake have no way to support currently. Don't suppose any of y'all have been secretly working to support this since April? 😄

**Comment by aaronbell** [comment]
One of the problems in the case of corner components, they are intended to be open paths:

<img width="549" alt="Screen Shot 2020-11-02 at 3 51 33 PM" src="https://user-images.githubusercontent.com/8460297/97931800-4f835500-1d23-11eb-992c-180c23065fdc.png">

So when glyphsLib hits them, it errors out:
```
fontmake: Error: In 'sources/aoboshi.glyphs' -> 'master_ufo/aoboshi-B.designspace' -> 'master_ufo/aoboshi-B.ufo': Compiling UFO failed: Unhandled open contour
```

I realize that preserving these in UFO might be challenging (for round trip processes), but even having a mechanism to decompose them only the relevant glyphs would help with the build process. Is that possible? Or am I going to have to decompose them in the source?

**Comment by schriftgestalt** [comment]
The ufo export in Glyphs has a "Decompose smart stuff" option that will decompose corners.

**Comment by aaronbell** [comment]
Thanks @schriftgestalt. I’m trying to build straight from the glyphs source though (the original designer works in Glyphs) so would rather not add a manual intermediary step.

**Comment by davelab6** [comment]
Scripting the glyphapp export and building from there is an immediate
workaround, but it would be good to get this done, Aaron.

 Nikolas, do you think you could implement it if you had time?

**Comment by madig** [comment]
Possibly, but I realistically don't, as we don't use smart anything at Dama
right now.

**Comment by schriftgestalt** [comment]
Maybe you should ;)

**Comment by jpt** [comment]
hmm. this could keep barlow in glyphs (3), probably. was just looking to see what the glyphs 3 implementation of smart corners could do for jpt/barlow#10 and I think it should work

**Comment by aaronbell** [comment]
@jpt you can set up a build script to remote control glyphs and export, if you want to use Python to build the font. Here’s an example:
https://github.com/fontdasu/ShipporiAntique/blob/master/build.py

**Comment by jpt** [comment]
that's a nice script for workflow automation, but it's not really the same benefit as being able to build a glyphs file in an environment where glyphs isn't installed. e.g. if fontlab 7 were using glyphslib under the hood for its glyphs support (not sure if that is the case)

**Comment by justvanrossum** [comment]
FWIW, I observed:

If the Smart Glyph doesn't have masters at all combined axis extremes, it is unclear how Glyphs does the extrapolation: it sure isn't the same as varLib, but also not the same as MutatorMath.

Also: it matters (at least for this case) in which order the axes are defined in the Smart Glyph.

**Comment by schriftgestalt** [comment]
Glyphs used the same algorithm for the smart components as it uses for the static instance. It used to use something close to mutatormath but that couldn't do intermediate masters and had quite big problems with extrapolations.

**Comment by justinpenner** [comment]
Hi all, just popping in here to add my support for this feature request. I think anyone who is working in Glyphs and has learned how to use cap/corner components would use them on almost all serif typefaces – that's the strongest argument I can come up with for implementing support for smart/cap/corner/etc components in glyphsLib.

Resorting to a script that opens the Glyphs application and controls it via code is a pretty ugly workaround (huge thanks to @aaronbell for coming up with it, though), but I think it's the only workable solution right now for all Glyphs users to be able to use build scripts in their workflow. And as @jpt mentioned, it would be beneficial to have a command line tool that could build from Glyphs sources on systems that don't have the Glyphs application installed.

If there's anything I can do to help (supplying source files with components, explaining how I use components in my workflow, etc.) please let me know.

**Comment by aaronbell** [comment]
@justinpenner Thanks, but I copied that wholesale from an example Georg gave me. :) No way I was figuring that out on my own!

**Comment by madig** [comment]
There is still no manpower from our side to do anything about this, not even enough to review [what is already written](https://github.com/googlefonts/glyphsLib/pull/652).
